Add Signing in order to be able to publish packages to nuget

safern opened the pull request:
Add targets to be able to sign packages and its contents in order to be able to push experimental or preview packages to nuget.org

This is based on arcade's Microsoft.DotNet.SignTool package.

I will test the build definition before merging.

JohnTortugo commented on tools/signing/sign.targets:
```diff
@@ -0,0 +1,61 @@
+
+    <_DesktopMSBuildRequired Condition="'$(_DryRun)' != 'true'">true</_DesktopMSBuildRequired>
+
+    <_SigningTempDir Condition="'$(_SigningTempDir)' == ''">$(MSBuildProjectDirectory)\obj\</_SigningTempDir>
```
You may want to add a `tmp` directory here. A bunch of folder\files will be created during the signing.

JohnTortugo commented on tools/signing/sign.targets:
```diff
@@ -0,0 +1,61 @@
+    <_DryRun>true</_DryRun>
+    <_DryRun Condition="'$(BUILD_BUILDNUMBER)' != ''">false</_DryRun>
+    <_TestSign>false</_TestSign>
+    <_TestSign Condition="'$(SignType)' == 'test'">true</_TestSign>
```
Maybe you can change this to just use some logic around SignType

safern replied:
Could you elaborate? I thought that is what I was doing.

JohnTortugo replied:
I meant to remove the `_TestSign` property and just use `SignType` instead. But looking again I think that might just make things more complicated anyway..

ahsonkhan commented on scripts/package.ps1:
```diff
@@ -38,6 +40,16 @@ foreach ($file in [System.IO.Directory]::EnumerateFiles("$repoRoot\src", "*.cspr
+    & $dotnetExePath msbuild "$repoRoot\tools\sign.proj" /p:SignType="$SignType" /p:PackagesPath="$packagesPath"
+
+    if (!$?) {
+        exit $lastExitCode
```
Nit: Write an error message before exiting

`Write-Error "Failed to sign the packages"`

ahsonkhan commented on tools/dependencies.props:
```diff
@@ -12,5 +12,9 @@
+    <SnVersion>1.0.0</SnVersion>
+    <VSWhereVersion>2.5.2</VSWhereVersion>
+    <MicroBuildCoreVersion>0.2.0</MicroBuildCoreVersion>
+    <MicrosoftDotNetSignToolVersion>1.0.0-beta.19058.5</MicrosoftDotNetSignToolVersion>
```
Where are these versions coming from?

safern replied:
This versions are coming from the arcade repo.

MicrosoftDotNetSignToolVersion is built in arcade.
Sn and VSWhere I just took the versions that arcade SDK was using to do signing.

ahsonkhan commented on tools/signing/sign.targets:
```diff
@@ -0,0 +1,61 @@
+  <Import Project="..\dependencies.props" />
+  <PropertyGroup>
+    <NuGetPackageRoot>$(RestorePackagesPath)</NuGetPackageRoot>
+    <NuGetPackageRoot Condition="!HasTrailingSlash('$(NuGetPackageRoot)')">$(NuGetPackageRoot)\</NuGetPackageRoot>
```
Can you please explain why `HasTrailingSlash` check is required here?

safern replied:
Just to avoid having a double '\' in the path. I can remove it if desired.

ahsonkhan commented on tools/sign.proj:
```diff
@@ -0,0 +1,14 @@
+<Project DefaultTargets="Build">
```
Can this also go under the signing directory?

safern replied:
Yup.

safern commented on NuGet.Config:
```diff
@@ -8,12 +8,13 @@
   -->
   <config>
-    <add key="repositoryPath" value="..\packages" />
+    <add key="repositoryPath" value=".nuget\packages" />
```
Nope, but nuget should create it when restoring from what I know.

safern commented on scripts/package.ps1:
```diff
@@ -1,7 +1,9 @@
     [string]$Configuration="Debug",
     [string]$ApiKey,
-    [string]$BuildVersion=[System.DateTime]::Now.ToString('preview2-yyMMdd-1')
+    [string]$BuildVersion=[System.DateTime]::Now.ToString('preview2-yyMMdd-1'),
```
Sure, I can do that. Do we want experimental? Or beta?

ahsonkhan replied:
They used to just have the `e` prefix to denote experimental (for example: 0.1.0-e160310-1)

https://dotnet.myget.org/feed/dotnet-corefxlab/package/nuget/System.Drawing.Graphics

Changing it back to 'e' would be fine, imo (as long as it doesn't break the sort order and the newest ones show up before the ones listed as preview).

safern replied:
I thin preview would sort as latest (because of alphabetical ordering) so then beta would probably work better.

cc: @ericstj

ericstj replied:
If you move backwards in prerelease specifier you should bump all the package versions so that "latest" does the right thing.

safern commented:
@ahsonkhan I fixed the versioning and also added the ability to pass a filter to the signing targets so with that we could only sign a specific package if wanted. PTAL
